Modules/Registration/Metricsv4/include/itkANTSNeighborhoodCorrelationImageToImageMetricv4GetValueAndDerivativeThreader.hxx

```diff
@@ -113,14 +113,16 @@ ANTSNeighborhoodCorrelationImageToImageMetricv4GetValueAndDerivativeThreader<
                                                           const DomainType & domain,
                                                           const ThreadIdType threadId)
 {
-  /* call base method */
-  /* Store the casted pointer to avoid dynamic casting in tight loops. */
-  this->m_ANTSAssociate = dynamic_cast<TNeighborhoodCorrelationMetric *>(this->m_Associate);
-  if (this->m_ANTSAssociate == nullptr)
-  {
-    itkExceptionMacro("Dynamic casting of associate pointer failed.");
-  }
+  std::call_once(this->m_ANTSAssociateOnceFlag, [this]() {
+    /* Store the casted pointer to avoid dynamic casting in tight loops. */
+    this->m_ANTSAssociate = dynamic_cast<TNeighborhoodCorrelationMetric *>(this->m_Associate);
+    if (this->m_ANTSAssociate == nullptr)
+    {
+      itkExceptionMacro("Dynamic casting of associate pointer failed.");
+    }
+  });
```

Comment:
Nice that you found this potential race condition. 👍  I'm sorry I don't know the code either, and I find it hard to understand. 🤷 The `m_ANTSAssociateOnceFlag` apparently ensures that `m_ANTSAssociate = associate` is only executed once during the lifetime of this member variable. Which means that it cannot be reset to a different "associate" anymore, during its lifetime. Is that OK?

It appears that  `m_Associate` is a protected member of `DomainThreader`. It is assigned in `DomainThreader::Execute`:

https://github.com/InsightSoftwareConsortium/ITK/blob/f408b77a57095a44964187af840353809f3644b0/Modules/Core/Common/include/itkDomainThreader.hxx#L54-L57

So then I guess the question is: could this "threader" be executed multiple times for different "associates"?

Reply:
It _seems_ to me that a mutex (member variable)+ lock_guard variables (locally) would be more appropriate than call_once + once_flag, in this case. But again, I don't know the code either.

Reply:
But then again, it seems a bit "overdone" to have a separate data member (m_ANTSAssociate), just to store the result of `dynamic_cast<TNeighborhoodCorrelationMetric *>(Superclass::m_Associate)`, and then try this data member keep it up-to-date and thread-safe.

Reply:
The `this->m_ANTSAssociate` should be change to just a function level variable with the results of the dynamic cast? This seems like it may have been an "optimization" which is not worth the complexity, with some unexpected cost?

Reply:
@seanm Looks like `m_ANTSAssociate` was actually introduced one month before songgang's commit already! Commit 96144090c6cacc2f8750c2de1f40ff5cd8b8abe9 authored by Michael Stauffer and committed by Hans (@hjmjohnson) has: https://github.com/InsightSoftwareConsortium/ITK/blob/96144090c6cacc2f8750c2de1f40ff5cd8b8abe9/Modules/Registration/Metricsv4/include/itkANTSNeighborhoodCorrelationImageToImageMetricv4DenseGetValueAndDerivativeThreader.h#L135-L137

And its commit text says:
> Remove calls to dynamic_cast in image metric threaders. Replace with cached pointer to metric object. Speed gain of several %.

Now I wonder if that speed gain will still be there when thread safety is taken into account 🤷 . Moreover, new compilers may yield different performance results anyway!

Reply:
I've looked a little more...

The string `Store the casted pointer to avoid dynamic casting in tight loops` appears 7 times, so seems, at least at one time, people thought it was a useful optimization (or got copy-pasted :))

>So then I guess the question is: could this "threader" be executed multiple times for different "associates"?

I think the answer is no. The string `m_Associate` "only" appears 108 times, and almost always it's read, not written. I can't find a setter for the ivar in fact. Only set is in `DomainThreader<TDomainPartitioner, TAssociate>::Execute`.